blog/_posts/2019-07-14-multithreading.md

```diff
@@ -0,0 +1,548 @@
+that the performance of this code continues to scale if we add a third thread:
+
+```
+julia> b = copy(a); @time psort!(b);
+  1.511860 seconds (3.77 k allocations: 686.935 MiB, 6.45% gc time)
+```
+
+Thinking about this two-way decomposition algorithm running on three threads
+can make your head hurt a little!
+In our view, this helps underscore how "automatic" this interface makes
+parallelism feel.
+
+Let's try a different machine with more CPU cores:
+
+```
+$ for n in 1 2 4 8 16; do    JULIA_NUM_THREADS=$n ./julia psort.jl; done
+  2.958881 seconds (3.58 k allocations: 686.932 MiB, 4.71% gc time)
+  1.868720 seconds (3.77 k allocations: 686.935 MiB, 7.03% gc time)
+  1.222777 seconds (3.78 k allocations: 686.935 MiB, 9.14% gc time)
+  0.958517 seconds (3.79 k allocations: 686.935 MiB, 18.21% gc time)
+  0.836891 seconds (3.78 k allocations: 686.935 MiB, 21.10% gc time)
+```
+
+Notice that this speedup occurs despite the parallel code allocating
+*drastically* more memory than the standard routine.
+The allocations come from two sources: `Task` objects, and the `temp`
+arrays allocated on each call.
+The reference sorting routine re-uses a single temporary buffer among
+all recursive calls.
+Re-using the temporary array is more difficult with parallelism, but
+still possible---more on that a little later.
+
+## How to move to a parallel world
+
+During the 1.3 series the new thread runtime is considered to be in beta testing.
+An "official" version will appear in a later release, to give us time to settle
+on an API we can commit to for the long term.
+Here's what you need to know if you want to upgrade your code over this period.
+
+### Task scheduling and synchronization
+
+To aid compatibility, code will continue to run within a single thread by default.
+When tasks are launched using existing primitives (`schedule`, `@async`), they
+will run only within the thread that launches them.
+Similarly, a `Condition` object (used to signal tasks when events occur) can only
+be used by the thread that created it.
+Attempts to wait for or notify conditions from other threads will raise errors.
+Separate thread-safe condition variables have been added, and are available as
+`Threads.Condition`.
+This needs to be a separate type because thread-safe use of condition variables
+requires acquiring a lock.
+In Julia, the lock is bundled with the condition, so `lock` can simply be called
+on the condition itself:
+
+```
+lock(cond::Threads.Condition)
+while !ready
+    wait(cond)
+end
+unlock(cond)
+```
+
+As in previous versions, the standard lock to use to protect critical sections
+is `ReentrantLock`, which is now thread-safe (it was previously only used for
+synchronizing tasks).
+`Threads.SpinLock` is also available, to be used in rare circumstances where
+(1) only threads and not tasks need to be synchronized, and (2) you expect to
+hold the lock for a short time.
+`Semaphore` and `Event` are also available, completing the standard set of
```

Comment:
not really complete? there'd also be barrier, rwlocks, and a "once" I think in a "standard set"
```suggestion
The `Threads` module also provides `Semaphore` and `Event` types with their standard definition.
```